### PR TITLE
Adding Implementation for noex::InstanceAdmin::AsyncListInstances

### PR DESCRIPTION
**Pull request by roopak-qlogic**
Fixes issue #1262

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1333)

### CODE REVIEW

File: google/cloud/bigtable/internal/instance_admin.h [hunk 1]
```diff
@@ -114,6 +114,54 @@ class InstanceAdmin {
...
+   * @tparam Functor the type of the callback.
+   */
+  template <typename Functor,
+            typename std::enable_if<
+                google::cloud::internal::is_invocable<
+                    Functor, CompletionQueue&,
+                    google::bigtable::admin::v2::ListInstancesResponse&,
+                    grpc::Status&>::value,
+                int>::type valid_callback_type = 0>
+  void AsyncListInstances(CompletionQueue& cq, Functor&& callback) {
```
Comment (coryan) [inline google/cloud/bigtable/internal/instance_admin.h]: This is just making one call to fetch the instances, if you look at the blocking version:

https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/8a54c7520df824e1fcd7b73c4da7916f6be69820/google/cloud/bigtable/internal/instance_admin.cc#L32

they make multiple calls in a loop.  You need to unroll that loop into a class that keeps the application callback, and the accumulated state for all the responses, and then invokes the callback with the final accumulated result.

If you read #1262 I requested that you propose an API first, I think this should be:

```C++
  template <typename Functor,
            typename std::enable_if<
                google::cloud::internal::is_invocable<
                    Functor, CompletionQueue&,
                    std::vector<google::bigtable::admin::v2::Instance>&,
                    grpc::Status&>::value,
                int>::type valid_callback_type = 0>
  void AsyncListInstances(CompletionQueue& cq, Functor&& callback) {
```

**Comment by roopak-qlogic** [comment]
Started working on implementation of  the complete AsysncListInstances, with reference of AsyncSampleRowKeys,  which currently saved in some other feature branch (https://github.com/GoogleCloudPlatform/google-cloud-cpp/compare/master...roopak-qlogic:Issue-1262_AsyncListInstancesImpl?expand=1 ) ,
due to some merge conflict.
I will merge final code here, once code will complete.

So  in current code , I am facing issue, that how to handle page token, that is seems not used in AsyncSampleRowKeys,

**Comment by dopiera** [comment]
FWIW, #1477 has been submitted and
https://github.com/dopiera/google-cloud-cpp/commit/7b70cd7c6aa02ab8bc0daa0f5d55193ea0f96c84 has a preview of `AsyncListClusters`. This PR should probably be similar.

The aforementioned preview, doesn't have unit tests yet, hopefully I'll add them today and submit a PR.

**Comment by roopak-qlogic** [comment]
As this PR is old one, and need to resolve lots of merge conflict, So closing this PR  and submitting new PR with 
https://github.com/GoogleCloudPlatform/google-cloud-cpp/pull/1572

**Pull request closed without merging**